snorkel/labeling/apply/lf_applier.py

```diff
@@ -0,0 +1,44 @@
+from itertools import chain
+from typing import Any, Iterable, List, Tuple
+
+import scipy.sparse as sparse
+from tqdm import tqdm
+
+from snorkel.labeling.lf import LabelingFunction
+from snorkel.types import Example
+
+RowData = List[Tuple[int, int, int]]
+
+
+class BaseLFApplier:
+    def __init__(self, lfs: List[LabelingFunction]) -> None:
+        if not all(isinstance(f, LabelingFunction) for f in lfs):
+            raise ValueError("lfs must be a list of LabelingFunctions")
+        self._lfs = lfs
+
+    def _matrix_from_row_data(self, labels: List[RowData]) -> sparse.csr_matrix:
+        row, col, data = zip(*chain.from_iterable(labels))
+        n, m = len(labels), len(self._lfs)
+        return sparse.csr_matrix((data, (row, col)), shape=(n, m))
+
+    def apply(self, examples: Any, *args: Any, **kwargs: Any) -> sparse.csr_matrix:
+        raise NotImplementedError
+
+
+def apply_lfs_to_example(
+    example: Example, index: int, lfs: List[LabelingFunction]
+) -> RowData:
+    labels = []
+    for j, lf in enumerate(lfs):
+        y = lf(example)
+        if y != 0:
+            labels.append((index, j, y))
+    return labels
```

Comment:
What is the rationale for not just using e.g. [`sparse.lil_matrix`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.lil_matrix.html) as the internal representation?

Reply:
And why not just have this be a class method?

Reply:
You mean lil vs. csr? Does metal need csr or no? I just usually default to csr since it's the best known sparse matrix format. lil, coo, etc. are scipy land.

Reply:
Conversely: why have it be a class method?

Reply:
No I just mean that you're incrementally constructing a sparse matrix (but just in a custom format) and then converting it to a `csr_matrix`.  However scipy already has a sparse matrix format, `lil_matrix`, optimized for incremental construction, and then has a method for conversion to `csr_matrix` (or any other type).

Note: this _is_ a nit, but I think will be cleaner!

Reply:
Ok sg

Reply:
@ajratner I implemented the changes in the new version, but

1. the following snippet is now repeated 3 times due to the pandas.df.apply and pyspark.rdd.map APIs

```
for j, lf in enumerate(self._lfs):
    y = lf(x)
    if y != 0:
```

2. the CSR construction is now slower for all the appliers because `chain.from_iterable(...)` is faster than the loop-based `lil.tocsr()`

<img width="1100" alt="Screen Shot 2019-06-14 at 10 52 02 AM" src="https://user-images.githubusercontent.com/7783678/59517854-79b84680-8e92-11e9-9684-8f7f21f7d3cc.png">

Reply:
Okay the difference is small, but based on this analysis, seems fine to default to the way you had before- I stand corrected!